l2-contracts/contracts/ConsensusRegistry.sol

```diff
@@ -0,0 +1,478 @@
+        attestersCommit++;
+
+        uint256 len = pendingRemovals.length;
+        for (uint256 i = 0; i < len; ++i) {
+            PendingNodeRemoval storage entry = pendingRemovals[i];
+            entry.pendingAttestersCommit = false; // Attester is considered as removed.
+            if (!entry.pendingValidatorsCommit) {
+                finalizeNodeRemoval(entry.nodeOwner);
+            }
+        }
+
+        emit AttestersCommitted();
+    }
+
+    function ensureAttesterSnapshot(Node storage node) internal {
+        if (attestersCommit > node.attesterLastUpdateCommit) {
+            node.attesterSnapshot = node.attesterLatest;
+        }
+    }
+
+    function ensureValidatorSnapshot(Node storage node) internal {
+        if (validatorsCommit > node.validatorLastUpdateCommit) {
+            node.validatorSnapshot = node.validatorLatest;
+        }
+    }
+
+    function scheduleNodeRemoval(address _nodeOwner) internal {
+        pendingRemovals.push(PendingNodeRemoval({
+            nodeOwner: _nodeOwner,
+            pendingAttestersCommit: true,
+            pendingValidatorsCommit: true
+        }));
+    }
+
+    function finalizeNodeRemoval(address _nodeOwner) internal onlyOwner {
+        // Remove from array by swapping the last element (gas-efficient, not preserving order).
+        nodeOwners[nodeOwnerIdx(_nodeOwner)] = nodeOwners[nodeOwners.length - 1];
+        nodeOwners.pop();
+
+        delete nodes[_nodeOwner];
+
+        // TODO: remove entry from pendingRemovals
+    }
+
+    /// @notice Finds the index of a node owner in the `nodeOwners` array.
+    /// @dev Throws an error if the node owner is not found in the array.
+    /// @param _nodeOwner The address of the node's owner to find in the `nodeOwners` array.
+    /// @return The index of the node owner in the `nodeOwners` array.
+    function nodeOwnerIdx(address _nodeOwner) private view returns (uint256) {
+        uint256 len = nodeOwners.length;
+        for (uint256 i = 0; i < len; ++i) {
+            if (nodeOwners[i] == _nodeOwner) {
+                return i;
+            }
+        }
+        revert NodeOwnerNotFound();
+    }
+
+    /// @notice Verifies that a node owner exists in the registry.
+    /// @dev Throws an error if the node owner does not exist.
+    /// @param _nodeOwner The address of the node's owner to verify.
+    function verifyNodeOwnerExists(address _nodeOwner) private view {
+        BLS12_381PublicKey storage pubKey = nodes[_nodeOwner].validatorLatest.pubKey;
+        if (
+            pubKey.a == bytes32(0) &&
+            pubKey.b == bytes32(0) &&
+            pubKey.c == bytes32(0)
+        ) {
+            revert NodeOwnerDoesNotExist();
+        }
+    }
+
+    function verifyInputAddress(address _nodeOwner) private pure {
+        if (_nodeOwner == address(0)) {
+            revert InvalidInputNodeOwnerAddress();
+        }
+    }
+
+    function verifyInputBLS12_381PublicKey(BLS12_381PublicKey calldata _pubKey) private pure {
```

Comment:
We use `_` prefix naming convention for `internal`/`private` functions

Reply:
Added here: https://github.com/matter-labs/era-contracts/pull/555/commits/9137d355745ac3d08d2975fe3ed6e08dba8859a1